XMonad/Hooks/DynamicLog.hs

```diff
@@ -585,6 +589,100 @@ pprWindowSetXinerama ws = "[" ++ unwords onscreen ++ "] " ++ unwords offscreen
+                        }
+
+-- | Create an 'IconSet' from a 'String'
+appIcon :: String -> IconSet
+appIcon x = pure [baseIconSet x]
+
+-- | Dynamically shows 'WorkspaceId' based on the 'Window's inside the Workspace
+-- which 'Icon's to use are defined via an 'IconSet'
+-- to create an 'IconSet' make a query that returns a '[Icon]'
+-- for example
+--
+-- > icons :: IconSet
+-- > icons = composeAll 
+-- >           [ className =? "discord" --> appIcon "\xfb6e" 
+-- >           , className =? "Discord" --> appIcon "\xf268"
+-- >           , className =? "Firefox" --> appIcon "\63288"
+-- >           , className =? "Spotify" <||>  className =? "spotify" --> appIcon "阮"
+-- >            ]
+-- >
+-- >
+-- >
+-- then you can add the hook to your config
+--
+-- > xmonad $ def 
+-- >      {
+-- >      logHook = dynamicLogIconsWithPP icons xmobarPP <> myManageHook
+-- >      }
+-- >
+--
+dynamicLogIconsWithPP :: IconSet -- ^ The 'IconSet' to use
+                         -> PP -- ^ The 'PP' to alter
+                         -> X () -- ^ The resulting 'X' action
+dynamicLogIconsWithPP iconset pp = dynamicLogIconsConvert (def{ iconConfigIcons = iconset, iconConfigPP = pp}) >>= dynamicLogWithPP 
+
+-- | Datatype for expanded 'Icon' configurations
+data IconConfig = IconConfig 
+    { iconConfigIcons :: IconSet -- ^ The 'IconSet' to use
+    , iconConfigStack :: [String] -> String -- ^ The function to manage stacking of 'Icon's
+    , iconConfigPP :: PP -- ^ The 'PP' to alter
+    }
+
+
+instance Default IconConfig where
+    def = IconConfig 
+        { iconConfigIcons = mempty
+        , iconConfigStack = ("["<>) . (<>"]") . unwords 
+        , iconConfigPP = def
+        }
+
+
+-- | This is the same as 'dynamicLogIconsWithPP' but it takes a 'IconConfig'
+-- this allows you to manually customise the 'Icon's the stacking function and also your `PP`
+dynamicLogIconsConvert :: IconConfig -> X PP
+dynamicLogIconsConvert IconConfig{iconConfigIcons = is, iconConfigStack = stack, iconConfigPP = pp} = do
+    ws <- gets windowset   
+    let workspaces' = map S.workspace (S.current ws : S.visible ws) <> S.hidden ws
+    icons <- M.fromList . (maybeToList =<<) <$> mapM (getIcons is) workspaces' 
+    pure $ pp 
+        { ppCurrent =  ppSection ppCurrent nameSetCurrent icons
+        , ppVisible =  ppSection ppVisible nameSetVisible icons
+        , ppHidden = ppSection ppHidden nameSetHidden icons
+        , ppHiddenNoWindows =  ppSection ppHiddenNoWindows nameSetHiddenNoWindows icons
+        }
+    where ppSection pF f icons = pF pp . concatIcons f . iconLookup icons
+          iconLookup icons x = M.findWithDefault [baseIconSet x] x icons
+          concatIcons f y 
+                | length y > 1 = stack $ map f y
+                | otherwise = concatMap f y
+
+getIcons :: IconSet -> S.Workspace WorkspaceId l Window -> X (Maybe (WorkspaceId,[Icon]))
+getIcons is w = do
+    let windows'' stack' = [S.focus stack'] ++ S.up stack' ++ S.down stack'
```

Comment:
Depending on how important this specific order of the windows is, why not `StackSet.integrate`?